### DISCUSSION TITLE
Does the Ultralytics SDK perform any GPU pre/postprocessing for computer vision?

### DISCUSSION THREAD

**Discussion by tonyreina** [ultralytics/ultralytics#13782] [General]
As part of our computer vision pipeline we first use [cv.createBackgroundSubtractorMOG2](https://docs.opencv.org/3.4/de/de1/group__video__motion.html#ga2beb2dee7a073809ccec60f145b6b29c)() and then perform Yolo object detection.

Is there any way for me to have both of these run directly on the GPU rather than having background subtraction on the CPU and then a copy to the GPU for Yolo?

I'm planning to upgrade to RTX A6000 which includes GPUDirect RDMA from my Ximea camera. Would the Ultralytics Yolo be able to handle GPUDirect transfers?

**Comment by glenn-jocher** [comment]
@tonyreina hello!

Thank you for your question. The Ultralytics SDK primarily focuses on efficient model inference and training, leveraging GPU acceleration for these tasks. However, it does not natively handle GPU-based pre/postprocessing tasks such as background subtraction.

For your specific use case, where you want to run both background subtraction and YOLO object detection directly on the GPU, you might need to integrate custom GPU-based preprocessing steps into your pipeline. One approach could be to use CUDA or OpenCL to implement the background subtraction on the GPU, ensuring that the data remains on the GPU for subsequent YOLO inference, thereby avoiding unnecessary data transfers between the CPU and GPU.

Regarding GPUDirect RDMA, this technology allows direct memory access between GPUs and other devices, such as your Ximea camera, without involving the CPU. While Ultralytics YOLO itself does not directly manage GPUDirect transfers, you can certainly leverage this capability in your custom pipeline to optimize data flow and minimize latency. You would need to handle the integration of GPUDirect RDMA within your preprocessing steps before passing the data to the YOLO model for inference.

Here's a high-level outline of how you might structure your pipeline:

1. **Capture Image**: Use GPUDirect RDMA to transfer images from your Ximea camera directly to GPU memory.
2. **Background Subtraction**: Implement or use an existing CUDA/OpenCL-based background subtraction algorithm.
3. **YOLO Inference**: Pass the preprocessed data directly to the YOLO model for object detection.

If you need further assistance with integrating these components, the community and forums are great places to seek advice and share your progress.

Best of luck with your project, and feel free to reach out if you have any more questions! 😊
